File: webstreampro/src/utils/excelParser.js
const XLSX = require('xlsx');

function parseExcelFile(filePath) {
  try {
    const workbook = XLSX.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    const worksheet = workbook.Sheets[sheetName];
    const data = XLSX.utils.sheet_to_json(worksheet);

    const broadcasts = data.map((row, index) => {
      const tags = row.tags ? 
        (typeof row.tags === 'string' ? row.tags.split(',').map(t => t.trim()) : []) : 
        [];

      return {
        title: row.title || `Untitled Broadcast ${index + 1}`,
        description: row.description || '',
        tags: tags,
        categoryId: row.categoryId ? String(row.categoryId) : '20',
        privacyStatus: row.privacyStatus || 'public',
        scheduledStartDate: row.scheduledStartDate || '',
        scheduledStartTime: row.scheduledStartTime || '',
        scheduledStartTime: formatScheduledTime(
          row.scheduledStartDate,
          row.scheduledStartTime
        ),
        thumbnailPath: row.thumbnailPath || '',
        streamId: row.streamId || '',
        streamKey: row.streamKey || '',
        latency: row.latency || 'normal',
        enableDvr: row.enableDvr !== false && row.enableDvr !== 'FALSE',
        enableEmbed: row.enableEmbed !== false && row.enableEmbed !== 'FALSE',
        recordFromStart: row.recordFromStart !== false && row.recordFromStart !== 'FALSE',
        madeForKids: row.madeForKids === true || row.madeForKids === 'TRUE',
        containsSyntheticMedia: row.containsSyntheticMedia === true || row.containsSyntheticMedia === 'TRUE',
        enableMonetization: row.enableMonetization === true || row.enableMonetization === 'TRUE'
      };
    });

    return { success: true, broadcasts };
  } catch (error) {
    console.error('Error parsing Excel file:', error);
    return { success: false, error: error.message };
  }
}

function formatScheduledTime(date, time) {
  if (!date || !time) {
    const now = new Date();
    now.setHours(now.getHours() + 1);
    return now.toISOString();
  }

  try {
    const dateStr = typeof date === 'string' ? date : excelDateToJSDate(date);
    const timeStr = typeof time === 'string' ? time : excelTimeToString(time);
    
    const [year, month, day] = dateStr.split('-');
    const [hour, minute] = timeStr.split(':');
    
    const scheduledDate = new Date(year, month - 1, day, hour, minute);
    return scheduledDate.toISOString();
  } catch (error) {
    console.error('Error formatting scheduled time:', error);
    const now = new Date();
    now.setHours(now.getHours() + 1);
    return now.toISOString();
  }
}

function excelDateToJSDate(excelDate) {
  const date = XLSX.SSF.parse_date_code(excelDate);
  return `${date.y}-${String(date.m).padStart(2, '0')}-${String(date.d).padStart(2, '0')}`;
}

function excelTimeToString(excelTime) {
  const totalMinutes = Math.round(excelTime * 24 * 60);
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`;
}

module.exports = { parseExcelFile };

File: headless-server/excel_parser.py
import pandas as pd
from datetime import datetime, timedelta

def parse_excel_file(file_path, default_date=None, default_time=None):
    """Parse Excel file for broadcast data"""
    try:
        df = pd.read_excel(file_path)
        
        broadcasts = []
        
        for index, row in df.iterrows():
            # Parse tags
            tags = []
            if pd.notna(row.get('tags')):
                tags = [tag.strip() for tag in str(row['tags']).split(',')]
            
            # Parse scheduled time
            scheduled_time = parse_scheduled_time(
                row.get('scheduledStartDate'),
                row.get('scheduledStartTime'),
                default_date,
                default_time
            )
            
            broadcast = {
                'title': row.get('title', f'Untitled Broadcast {index + 1}'),
                'description': row.get('description', ''),
                'tags': tags,
                'categoryId': str(row.get('categoryId', '20')),
                'privacyStatus': row.get('privacyStatus', 'public'),
                'scheduledStartTime': scheduled_time,
                'thumbnailPath': row.get('thumbnailPath', ''),
                'streamId': row.get('streamId', ''),
                'streamKey': row.get('streamKey', ''),
                'latency': row.get('latency', 'normal'),
                'enableDvr': bool(row.get('enableDvr', True)),
                'enableEmbed': bool(row.get('enableEmbed', True)),
                'recordFromStart': bool(row.get('recordFromStart', True)),
                'madeForKids': bool(row.get('madeForKids', False)),
                'containsSyntheticMedia': bool(row.get('containsSyntheticMedia', False)),
                'enableMonetization': bool(row.get('enableMonetization', False))
            }
            
            broadcasts.append(broadcast)
        
        return True, broadcasts
    except Exception as e:
        return False, str(e)

def parse_scheduled_time(date_val, time_val, default_date=None, default_time=None):
    """Parse scheduled time from date and time values"""
    try:
        # Use defaults if not provided
        if pd.isna(date_val):
            if default_date:
                date_val = default_date
            else:
                # Default to tomorrow
                date_val = (datetime.now() + timedelta(days=1)).date()
        
        if pd.isna(time_val):
            if default_time:
                time_val = default_time
            else:
                # Default to 12:00
                time_val = "12:00"
        
        # Parse date
        if isinstance(date_val, str):
            date_obj = datetime.strptime(date_val, '%Y-%m-%d').date()
        else:
            date_obj = date_val if hasattr(date_val, 'date') else date_val
        
        # Parse time
        if isinstance(time_val, str):
            time_obj = datetime.strptime(time_val, '%H:%M').time()
        else:
            time_obj = time_val if hasattr(time_val, 'time') else time_val
        
        # Combine
        scheduled_dt = datetime.combine(date_obj, time_obj)
        
        return scheduled_dt.isoformat()
    except Exception as e:
        print(f"Error parsing scheduled time: {e}")
        # Default to 1 hour from now
        default_time = datetime.now() + timedelta(hours=1)
        return default_time.isoformat()
